quic/s2n-quic-platform/src/io/tokio/task/unix.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    features::Gso,
+    socket::{
+        ring,
+        task::{rx, tx},
+    },
+    syscall::{SocketType, UnixMessage},
+};
+use core::task::{Context, Poll};
+use std::{io, os::unix::io::AsRawFd};
+use tokio::io::unix::AsyncFd;
+
+pub async fn rx<S: Into<std::net::UdpSocket>, M: UnixMessage + Unpin>(
+    socket: S,
+    producer: ring::Producer<M>,
+) -> io::Result<()> {
+    let socket = socket.into();
+    socket.set_nonblocking(true).unwrap();
+
+    let socket = AsyncFd::new(socket).unwrap();
+    let result = rx::Receiver::new(producer, socket).await;
+    if let Some(err) = result {
+        Err(err)
+    } else {
+        Ok(())
+    }
+}
+
+pub async fn tx<S: Into<std::net::UdpSocket>, M: UnixMessage + Unpin>(
+    socket: S,
+    consumer: ring::Consumer<M>,
+    gso: Gso,
+) -> io::Result<()> {
+    let socket = socket.into();
+    socket.set_nonblocking(true).unwrap();
+
+    let socket = AsyncFd::new(socket).unwrap();
+    let result = tx::Sender::new(consumer, socket, gso).await;
+    if let Some(err) = result {
+        Err(err)
+    } else {
+        Ok(())
+    }
+}
+
+impl<S: AsRawFd, M: UnixMessage> tx::Socket<M> for AsyncFd<S> {
+    type Error = io::Error;
+
+    #[inline]
+    fn send(
+        &mut self,
+        cx: &mut Context,
+        entries: &mut [M],
+        events: &mut tx::Events,
+    ) -> io::Result<()> {
+        M::send(self.get_ref().as_raw_fd(), entries, events);
+
+        if !events.is_blocked() {
+            return Ok(());
+        }
+
+        for i in 0..2 {
+            match self.poll_write_ready(cx) {
```

Comment:
Added a comment; let me know if that makes sense

Reply:
Yes, I think this makes sense. I can't say I'm fully convinced that retrieving the guard after we call into I/O works in all cases - Tokio definitely seems to want you go "guard, then IO, then mark as complete if WouldBlock" - but I'm not sure why and it may not matter. Semantically this seems fine to me at a high level.